Fix a race when using TextMarshal on a StdDuration

dtjm opened the pull request:
- Benchmarks show performance to be roughly the same when performing the copy, no additional bytes allocated and no additional number of allocations
- ~~~Add in sync.Pool for bufio.Writer. Reduced allocations by 2 and run time by about 30% for TextMarshaler.Marshal~~~  (move this to upstream protobuf)

@awalterschulze I didn't see a guide for contributing. Please let me know if you accept pull requests and what needs to be done to get it approved.

Thanks,
/sam

awalterschulze commented:
Interesting.  I am on holiday.
I will look at this urgently when I get back in a week and a half.

awalterschulze commented on test/stdtypes/concurrency_test.go:
```diff
@@ -0,0 +1,51 @@
+	"github.com/gogo/protobuf/proto"
+)
+
+func TestConcurrentTextMarshal(t *testing.T) {
```
Does this test fail when you run it with -test.race ?
Or how does this test fail.

If it requires -test.race then maybe we should make a Makefile that actually calls this with -test.race to make sure we don't revert this change in future.

dtjm replied:
The test requires -race to trigger the race condition.  Running this test against master yields:

```
# No race detector
$ go test ./test/stdtypes/
ok  	github.com/gogo/protobuf/test/stdtypes	0.032s

# With race detector
$ go test -race ./test/stdtypes/
==================
WARNING: DATA RACE
Read at 0x00c4200cc358 by goroutine 8:
  github.com/gogo/protobuf/proto.(*TextMarshaler).writeAny()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:525 +0xc4d
  github.com/gogo/protobuf/proto.(*TextMarshaler).writeStruct()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:461 +0x13e1
  github.com/gogo/protobuf/proto.(*TextMarshaler).Marshal()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:894 +0x4aa
  github.com/gogo/protobuf/test/stdtypes.TestConcurrentTextMarshal.func2()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/test/stdtypes/concurrent_test.go:33 +0xa7

Previous write at 0x00c4200cc358 by goroutine 7:
  github.com/gogo/protobuf/proto.(*TextMarshaler).writeAny()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:534 +0xd3d
  github.com/gogo/protobuf/proto.(*TextMarshaler).writeStruct()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:461 +0x13e1
  github.com/gogo/protobuf/proto.(*TextMarshaler).Marshal()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:894 +0x4aa
  github.com/gogo/protobuf/test/stdtypes.TestConcurrentTextMarshal.func1()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/test/stdtypes/concurrent_test.go:24 +0xa7

Goroutine 8 (running) created at:
  github.com/gogo/protobuf/test/stdtypes.TestConcurrentTextMarshal()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/test/stdtypes/concurrent_test.go:37 +0x29c
  testing.tRunner()
      /usr/local/Cellar/go/1.8.1/libexec/src/testing/testing.go:657 +0x107

Goroutine 7 (running) created at:
  github.com/gogo/protobuf/test/stdtypes.TestConcurrentTextMarshal()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/test/stdtypes/concurrent_test.go:28 +0x242
  testing.tRunner()
      /usr/local/Cellar/go/1.8.1/libexec/src/testing/testing.go:657 +0x107
==================
==================
WARNING: DATA RACE
Read at 0x00c4200cc4b9 by goroutine 8:
  github.com/gogo/protobuf/proto.(*TextMarshaler).writeAny()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:538 +0xf66
  github.com/gogo/protobuf/proto.(*TextMarshaler).writeStruct()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:461 +0x13e1
  github.com/gogo/protobuf/proto.(*TextMarshaler).Marshal()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:894 +0x4aa
  github.com/gogo/protobuf/test/stdtypes.TestConcurrentTextMarshal.func2()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/test/stdtypes/concurrent_test.go:33 +0xa7

Previous write at 0x00c4200cc4b9 by goroutine 7:
  github.com/gogo/protobuf/proto.(*TextMarshaler).writeAny()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:546 +0x1104
  github.com/gogo/protobuf/proto.(*TextMarshaler).writeStruct()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:461 +0x13e1
  github.com/gogo/protobuf/proto.(*TextMarshaler).Marshal()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:894 +0x4aa
  github.com/gogo/protobuf/test/stdtypes.TestConcurrentTextMarshal.func1()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/test/stdtypes/concurrent_test.go:24 +0xa7

Goroutine 8 (running) created at:
  github.com/gogo/protobuf/test/stdtypes.TestConcurrentTextMarshal()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/test/stdtypes/concurrent_test.go:37 +0x29c
  testing.tRunner()
      /usr/local/Cellar/go/1.8.1/libexec/src/testing/testing.go:657 +0x107

Goroutine 7 (finished) created at:
  github.com/gogo/protobuf/test/stdtypes.TestConcurrentTextMarshal()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/test/stdtypes/concurrent_test.go:28 +0x242
  testing.tRunner()
      /usr/local/Cellar/go/1.8.1/libexec/src/testing/testing.go:657 +0x107
==================
--- FAIL: TestConcurrentTextMarshal (0.00s)
	testing.go:610: race detected during execution of test
FAIL
FAIL	github.com/gogo/protobuf/test/stdtypes	0.061s
```

awalterschulze replied:
Ok Excellent can we make sure that this one test is run with the race flag when we run the travis job.

dtjm replied:
@awalterschulze can you give me a hint on how to run this one test with the -race flag? It looks like all tests are run under one `go test ./...` invocation.  Should I add it to the individual `test/stdtypes/Makefile`?

dtjm replied:
I just added it to the `test/stdtypes/Makefile` for now. Let me know if that works.

awalterschulze replied:
Sorry I was sleeping, time difference :)

awalterschulze commented on proto/text.go:
```diff
@@ -856,6 +856,22 @@ type TextMarshaler struct {
 	ExpandAny bool // expand google.protobuf.Any messages of known types
 }
 
+var bufWriterPool = sync.Pool{
```
This bufWriterPool is a separate optimization.
I think we can try to contribute this to golang/protobuf and then merge this into the gogo/protobuf fork, when merged into golang/protobuf.

dtjm replied:
sounds good, I can move this change to the upstream golang/protobuf project

awalterschulze commented on test/stdtypes/concurrency_test.go:
```diff
@@ -0,0 +1,51 @@
+)
+
+func TestConcurrentTextMarshal(t *testing.T) {
+	t.Log(`Verify that there are no race conditions when calling
```
I don't think we need this as a log, we only need a comment.

dtjm replied:
will change to comment

awalterschulze commented on test/stdtypes/concurrency_test.go:
```diff
@@ -0,0 +1,51 @@
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			t.Logf("%v", std)
```
Do we need this log?

dtjm replied:
nope, it was just for debugging

awalterschulze commented on test/stdtypes/Makefile:
```diff
@@ -34,4 +34,5 @@ regenerate:
 	Mgoogle/protobuf/timestamp.proto=github.com/gogo/protobuf/types\
 	:. \
 	--proto_path=../../../../../:../../protobuf/:. stdtypes.proto
+	go test -race .
```
Lets add a separate test target to this Makefile 
```
test:
  go test -race -run= TestConcurrentTextMarshal .
```

And then we can add this to the top Makefile
https://github.com/gogo/protobuf/blob/master/Makefile#L117-L119

```
tests:
  go build ./test/enumprefix
  go test ./...
  (cd test/stdtypes && make test)
```

awalterschulze commented:
Remember to add yourself as a contributor or author as appropriate.

dtjm commented:
@awalterschulze I was able to get approval from our legal dept. I've squashed the commits together and rebased on top of the latest master. Please let me know if there's anything else I need to do here before it's ok to merge. Thanks!

dtjm commented on CONTRIBUTORS:
```diff
@@ -12,9 +12,11 @@ John Tuley <john@tuley.org>
 Tamir Duberstein <tamird@gmail.com>
 Todd Eisenberger <teisenberger@dropbox.com>
 Tormod Erevik Lea <tormodlea@gmail.com>
+Vyacheslav Kim <kane@sendgrid.com>
```
@Kane-Sendgrid FYI

Kane-Sendgrid replied:
@dtjm yes, thanks!

awalterschulze commented:
Maybe you also want to add sendgrid to the list of users in the Readme.md?

dtjm commented:
@awalterschulze good idea, added.

awalterschulze commented:
Perfect :)

awalterschulze commented:
Thanks again for the fix.  I really appreciate it.

dtjm commented:
Thank you @awalterschulze for your work on this project.
